# PyTorch Basics: Tensors & Gradients

## Tensors

In [1]:
import torch

At its core, PyTorch is a library for processing tensors. A tensor is a number, vector, matrix, or any n-dimensional array. Let's create a tensor with a single number.

In [2]:
# Number
t1 = torch.tensor(4.)
t1

tensor(4.)

`4.` is a shorthand for 4.0. It is used to indicate to Python (and PyTorch) that you want to create a floating-point number. We can verify this by checking the dtype attribute of our tensor.

In [3]:
t1.type

<function Tensor.type>

Let's try creating more complex tensors.

In [5]:
t2=torch.tensor([1.,2.,3.,4.])
t2

tensor([1., 2., 3., 4.])

In [6]:
# Matrix
t3 = torch.tensor([[5., 6], 
                   [7, 8], 
                   [9, 10]])
t3

tensor([[ 5.,  6.],
        [ 7.,  8.],
        [ 9., 10.]])

In [7]:
# 3-dimensional array
t4 = torch.tensor([
    [[11, 12, 13], 
     [13, 14, 15]], 
    [[15, 16, 17], 
     [17, 18, 19.]]])
t4

tensor([[[11., 12., 13.],
         [13., 14., 15.]],

        [[15., 16., 17.],
         [17., 18., 19.]]])

Tensors can have any number of dimensions and different lengths along each dimension. We can inspect the length along each dimension using the .shape property of a tensor.

In [8]:
print(t1)
t1.shape

tensor(4.)


torch.Size([])

In [9]:
print(t2)
t2.shape

tensor([1., 2., 3., 4.])


torch.Size([4])

In [10]:
print(t3)
t3.shape

tensor([[ 5.,  6.],
        [ 7.,  8.],
        [ 9., 10.]])


torch.Size([3, 2])

In [11]:
print(t4)
t4.shape

tensor([[[11., 12., 13.],
         [13., 14., 15.]],

        [[15., 16., 17.],
         [17., 18., 19.]]])


torch.Size([2, 2, 3])

Note that it's not possible to create tensors with an improper shape.



## Tensor operations and gradients


We can combine tensors with the usual arithmetic operations. Let's look at an example:



In [12]:
x = torch.tensor(3.)
w = torch.tensor(4., requires_grad=True)
b = torch.tensor(5., requires_grad=True)
x, w, b

(tensor(3.), tensor(4., requires_grad=True), tensor(5., requires_grad=True))

We've created three tensors: `x`,` w,` and `b,` all numbers. w and b have an additional parameter `requires_grad` set to `True.` We'll see what it does in just a moment.

Let's create a new tensor y by combining these tensors.

In [13]:
y=w*x+b
y

tensor(17., grad_fn=<AddBackward0>)

As expected, y is a tensor with the value 3 * 4 + 5 = 17. What makes PyTorch unique is that we can automatically compute the derivative of y w.r.t. the tensors that have requires_grad set to True i.e. w and b. This feature of PyTorch is called autograd (automatic gradients).

To compute the derivatives, we can invoke the .backward method on our result y.

In [14]:
y.backward()

The derivatives of y with respect to the input tensors are stored in the .grad property of the respective tensors.

In [15]:
# Display gradients
print('dy/dx:', x.grad)
print('dy/dw:', w.grad)
print('dy/db:', b.grad)

dy/dx: None
dy/dw: tensor(3.)
dy/db: tensor(1.)


**Tensor functions**


Apart from arithmetic operations, the torch module also contains many functions for creating and manipulating tensors. Let's look at some examples.



In [16]:
# Create a tensor with a fixed value for every element
t6 = torch.full((3, 2), 42)
t6

tensor([[42, 42],
        [42, 42],
        [42, 42]])

In [17]:
# Concatenate two tensors with compatible shapes
t7 = torch.cat((t3, t6))
t7

tensor([[ 5.,  6.],
        [ 7.,  8.],
        [ 9., 10.],
        [42., 42.],
        [42., 42.],
        [42., 42.]])

In [18]:
# Compute the sin of each element
t8 = torch.sin(t7)
t8

tensor([[-0.9589, -0.2794],
        [ 0.6570,  0.9894],
        [ 0.4121, -0.5440],
        [-0.9165, -0.9165],
        [-0.9165, -0.9165],
        [-0.9165, -0.9165]])

In [19]:
# Compute the cos of each element
t8 = torch.cos(t7)
t8

tensor([[ 0.2837,  0.9602],
        [ 0.7539, -0.1455],
        [-0.9111, -0.8391],
        [-0.4000, -0.4000],
        [-0.4000, -0.4000],
        [-0.4000, -0.4000]])

In [20]:
# Change the shape of a tensor
t9 = t8.reshape(3, 2, 2)
t9

tensor([[[ 0.2837,  0.9602],
         [ 0.7539, -0.1455]],

        [[-0.9111, -0.8391],
         [-0.4000, -0.4000]],

        [[-0.4000, -0.4000],
         [-0.4000, -0.4000]]])

**Interoperability with Numpy**

In [25]:
import numpy as np

x = np.array([[1, 2], [3, 4.]])
x

array([[1., 2.],
       [3., 4.]])

We can convert a Numpy array to a PyTorch tensor using `torch.from_numpy`.

In [22]:
# Convert the numpy array to a torch tensor.
y = torch.from_numpy(x)
y

tensor([[1., 2.],
        [3., 4.]], dtype=torch.float64)

In [26]:
x.dtype, y.dtype

(dtype('float64'), torch.float64)

We can convert a PyTorch tensor to a Numpy array using the `.numpy` method of a tensor.



In [27]:
# Convert a torch tensor to a numpy array
z = y.numpy()
z

array([[1., 2.],
       [3., 4.]])

The interoperability between PyTorch and Numpy is essential because most datasets you'll work with will likely be read and preprocessed as Numpy arrays.

You might wonder why we need a library like PyTorch at all since Numpy already provides data structures and utilities for working with multi-dimensional numeric data. There are two main reasons:

Autograd: The ability to automatically compute gradients for tensor operations is essential for training deep learning models.
GPU support: While working with massive datasets and large models, PyTorch tensor operations can be performed efficiently using a Graphics Processing Unit (GPU). Computations that might typically take hours can be completed within minutes using GPUs.
We'll leverage both these features of PyTorch extensively in this tutorial series.

## Gradient Descent and Linear Regression with PyTorch


**Simple regression**

In this tutorial, we'll discuss one of the foundational algorithms in machine learning: Linear regression. We'll create a model that predicts crop yields for apples and oranges (target variables) by looking at the average temperature, rainfall, and humidity (input variables or features) in a region. Here's the training data:

We can represent the training data using two matrices: inputs and targets, each with one row per observation, and one column per variable

In [34]:
# Input (temp, rainfall, humidity)
inputs = np.array([[73, 67, 43], 
                   [91, 88, 64], 
                   [87, 134, 58], 
                   [102, 43, 37], 
                   [69, 96, 70]], dtype='float32')

In [35]:
# Targets (apples, oranges)
targets = np.array([[56, 70], 
                    [81, 101], 
                    [119, 133], 
                    [22, 37], 
                    [103, 119]], dtype='float32')

We've separated the input and target variables because we'll operate on them separately. Also, we've created numpy arrays, because this is typically how you would work with training data: read some CSV files as numpy arrays, do some processing, and then convert them to PyTorch tensors.

Let's convert the arrays to PyTorch tensors

In [36]:
inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)
print(inputs)
print(targets)

tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.]])
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


The weights and biases `(w11, w12,... w23, b1 & b2)` can also be represented as matrices, initialized as random values. The first row of `w` and the first element of `b` are used to predict the first target variable, i.e., yield of apples, and similarly, the second for oranges

In [37]:
# Weights and biases
w = torch.randn(2, 3, requires_grad=True)
b = torch.randn(2, requires_grad=True)
print(w)
print(b)

tensor([[ 0.5121,  0.0860, -0.1805],
        [ 1.3327, -1.7127,  0.5562]], requires_grad=True)
tensor([-0.5312, -1.3984], requires_grad=True)


`torch.randn` creates a tensor with the given shape, with elements picked randomly from a normal distribution with mean 0 and standard deviation 1.

Our model is simply a function that performs a matrix multiplication of the `inputs` and the weights `w` (transposed) and adds the bias `b` (replicated for each observation).

We can define the model as follows:



In [38]:
def model(x):
    return x @ w.t() + b

`@`represents matrix multiplication in PyTorch, and the `.t` method returns the transpose of a tensor.

The matrix obtained by passing the input data into the model is a set of predictions for the target variables.

In [39]:
# Generate predictions
preds=model(inputs)
print(preds)

tensor([[ 34.8536,   5.0537],
        [ 42.0872,   4.7558],
        [ 45.0799, -82.6961],
        [ 48.7221,  81.4687],
        [ 30.4267, -34.9273]], grad_fn=<AddBackward0>)


Let's compare the predictions of our model with the actual targets.



In [40]:
print(targets)

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


You can see a big difference between our model's predictions and the actual targets because we've initialized our model with random weights and biases. Obviously, we can't expect a randomly initialized model to just work.

**Loss function**

Before we improve our model, we need a way to evaluate how well our model is performing. We can compare the model's predictions with the actual targets using the following method:

- Calculate the difference between the two matrices (preds and targets).
- Square all elements of the difference matrix to remove negative values.
- Calculate the average of the elements in the resulting matrix.

The result is a single number, known as the `mean squared error (MSE)`.

In [41]:
# MSE loss
def mse(t1, t2):
    diff = t1 - t2
    return torch.sum(diff * diff) / diff.numel()

`torch.sum` returns the sum of all the elements in a tensor. The `.numel` method of a tensor returns the number of elements in a tensor. Let's compute the mean squared error for the current predictions of our model.

In [42]:
# Compute loss
loss=mse(preds, targets)
print(loss)

tensor(9908.3359, grad_fn=<DivBackward0>)


Here’s how we can interpret the result: On average, each element in the prediction differs from the actual target by the square root of the loss. And that’s pretty bad, considering the numbers we are trying to predict are themselves in the range 50–200. The result is called the loss because it indicates how bad the model is at predicting the target variables. It represents information loss in the model: *the lower the loss, the better the model.*

**Compute gradients**


With PyTorch, we can automatically compute the gradient or derivative of the loss `w.r.t`. to the weights and biases because they have `requires_grad` set to `True`. We'll see how this is useful in just a moment.

In [43]:
# Compute gradients
loss.backward()

The gradients are stored in the `.grad` property of the respective tensors. Note that the derivative of the loss w.r.t. the weights matrix is itself a matrix with the same dimensions.

In [45]:
print(w.grad)
print(w)

tensor([[ -2759.5403,  -4112.8838,  -2355.6990],
        [ -7670.0073, -10917.8066,  -6118.4521]])
tensor([[ 0.5121,  0.0860, -0.1805],
        [ 1.3327, -1.7127,  0.5562]], requires_grad=True)


**Adjust weights and biases to reduce the loss**

Adjust weights and biases to reduce the loss
The loss is a quadratic function of our weights and biases, and our objective is to find the set of weights where the loss is the lowest. If we plot a graph of the loss w.r.t any individual weight or bias element, it will look like the figure shown below. An important insight from calculus is that the gradient indicates the rate of change of the loss, i.e., the loss function's slope w.r.t. the weights and biases.

If a gradient element is positive:

- increasing the weight element's value slightly will increase the loss
- decreasing the weight element's value slightly will decrease the loss


If a gradient element is negative:

- increasing the weight element's value slightly will decrease the loss
- decreasing the weight element's value slightly will increase the loss


The increase or decrease in the loss by changing a weight element is proportional to the gradient of the loss w.r.t. that element. This observation forms the basis of the **gradient descent optimization** algorithm that we'll use to improve our model **(by descending along the gradient)**.

We can subtract from each weight element a small quantity proportional to the derivative of the loss w.r.t. that element to reduce the loss slightly.

In [46]:
w
w.grad

tensor([[ -2759.5403,  -4112.8838,  -2355.6990],
        [ -7670.0073, -10917.8066,  -6118.4521]])

In [47]:
with torch.no_grad():
    w -= w.grad * 1e-5
    b -= b.grad * 1e-5

We multiply the gradients with a very small number (10^-5 in this case) to ensure that we don't modify the weights by a very large amount. We want to take a small step in the downhill direction of the gradient, not a giant leap. This number is called the **learning rate** of the algorithm.

We use `torch.no_grad` to indicate to PyTorch that we shouldn't track, calculate, or modify gradients while updating the weights and biases.

In [48]:
# Let's verify that the loss is actually lower
loss = mse(preds, targets)
print(loss)

tensor(9908.3359, grad_fn=<DivBackward0>)


Before we proceed, we reset the gradients to zero by invoking the `.zero_()` method. We need to do this because PyTorch accumulates gradients. Otherwise, the next time we invoke `.backward` on the loss, the new gradient values are added to the existing gradients, which may lead to unexpected results.

In [49]:
w.grad.zero_()
b.grad.zero_()
print(w.grad)
print(b.grad)

tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([0., 0.])


**Train the model using gradient descent**


As seen above, we reduce the loss and improve our model using the gradient descent optimization algorithm. Thus, we can train the model using the following steps:

1. Generate predictions

2. Calculate the loss

3. Compute gradients w.r.t the weights and biases

4. Adjust the weights by subtracting a small quantity proportional to the gradient

5. Reset the gradients to zero

Let's implement the above step by step.

In [51]:
# 1. Generate predictions
preds = model(inputs)
print(preds)

tensor([[ 40.6370,  20.5997],
        [ 49.7258,  25.2600],
        [ 54.3586, -57.8436],
        [ 54.1774,  96.2515],
        [ 37.9285, -14.8700]], grad_fn=<AddBackward0>)


In [52]:
# 2. Calculate the loss
loss = mse(preds, targets)
print(loss)

tensor(7669.2451, grad_fn=<DivBackward0>)


In [53]:
# 3. Compute gradients
loss.backward()
print(w.grad)
print(b.grad)

tensor([[-2159.8198, -3461.3286, -1955.1613],
        [-6059.0674, -9170.3398, -5043.8203]])
tensor([-28.8346, -78.1205])


In [57]:
# 4. Adjust weights & reset gradients
with torch.no_grad():
    w -= w.grad * 1e-5
    b -= b.grad * 1e-5
    w.grad.zero_()
    b.grad.zero_()

In [58]:
print(w)
print(b)

tensor([[ 0.5613,  0.1618, -0.1374],
        [ 1.4700, -1.5118,  0.6678]], requires_grad=True)
tensor([-0.5305, -1.3967], requires_grad=True)


With the new weights and biases, the model should have a lower loss.

In [59]:
# Calculate loss
preds = model(inputs)
loss = mse(preds, targets)
print(loss)

tensor(6147.9209, grad_fn=<DivBackward0>)


We have already achieved a significant reduction in the loss merely by adjusting the weights and biases slightly using gradient descent.



**Train for multiple epochs**


To reduce the loss further, we can repeat the process of adjusting the weights and biases using the gradients multiple times. Each iteration is called an epoch. Let's train the model for 100 epochs.

In [61]:
for i in range(100):
    preds=model(inputs)
    loss=mse(preds,targets)
    loss.backward()
    with torch.no_grad():
        w -= w.grad * 1e-5
        b -= b.grad * 1e-5
        w.grad.zero_()
        b.grad.zero_()

Once again, let's verify that the loss is now lower:



In [62]:
# Calculate loss
preds = model(inputs)
loss = mse(preds, targets)
print(loss)

tensor(880.3444, grad_fn=<DivBackward0>)


The loss is now much lower than its initial value. Let's look at the model's predictions and compare them with the targets.



In [63]:
# Predictions
preds

tensor([[ 62.7512,  81.8202],
        [ 81.7975, 111.3481],
        [110.6420,  90.1632],
        [ 53.6441, 105.5727],
        [ 82.1010,  97.4305]], grad_fn=<AddBackward0>)

In [65]:
# Targets
targets

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])

The predictions are now quite close to the target variables. We can get even better results by training for a few more epochs.



**Linear regression using PyTorch built-ins**

We've implemented linear regression & gradient descent model using some basic tensor operations. However, since this is a common pattern in deep learning, PyTorch provides several built-in functions and classes to make it easy to create and train models with just a few lines of code.

Let's begin by importing the `torch.nn` package from PyTorch, which contains utility classes for building neural networks.

In [67]:
import torch.nn as nn

As before, we represent the inputs and targets and matrices.

In [68]:
# Input (temp, rainfall, humidity)
inputs = np.array([[73, 67, 43], 
                   [91, 88, 64], 
                   [87, 134, 58], 
                   [102, 43, 37], 
                   [69, 96, 70], 
                   [74, 66, 43], 
                   [91, 87, 65], 
                   [88, 134, 59], 
                   [101, 44, 37], 
                   [68, 96, 71], 
                   [73, 66, 44], 
                   [92, 87, 64], 
                   [87, 135, 57], 
                   [103, 43, 36], 
                   [68, 97, 70]], 
                  dtype='float32')

# Targets (apples, oranges)
targets = np.array([[56, 70], 
                    [81, 101], 
                    [119, 133], 
                    [22, 37], 
                    [103, 119],
                    [57, 69], 
                    [80, 102], 
                    [118, 132], 
                    [21, 38], 
                    [104, 118], 
                    [57, 69], 
                    [82, 100], 
                    [118, 134], 
                    [20, 38], 
                    [102, 120]], 
                   dtype='float32')

inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)

We are using 15 training examples to illustrate how to work with large datasets in small batches.



**Dataset and DataLoader**


We'll create a `TensorDataset`, which allows access to rows from `inputs` and `targets` as tuples, and provides standard APIs for working with many different types of datasets in PyTorch.

In [70]:
from torch.utils.data import TensorDataset

In [71]:
# Define dataset
train_ds = TensorDataset(inputs, targets)
train_ds[0:3]

(tensor([[ 73.,  67.,  43.],
         [ 91.,  88.,  64.],
         [ 87., 134.,  58.]]),
 tensor([[ 56.,  70.],
         [ 81., 101.],
         [119., 133.]]))

The `TensorDataset` allows us to access a small section of the training data using the array indexing notation ([0:3] in the above code). It returns a tuple with two elements. The first element contains the input variables for the selected rows, and the second contains the targets.

We'll also create a `DataLoader`, which can split the data into batches of a predefined size while training. It also provides other utilities like shuffling and random sampling of the data.

In [72]:
from torch.utils.data import DataLoader

In [73]:
# Define data loader
batch_size = 5
train_dl = DataLoader(train_ds, batch_size, shuffle=True)

We can use the data loader in a for loop. Let's look at an example.



In [74]:
for xb, yb in train_dl:
    print(xb)
    print(yb)
    break

tensor([[ 73.,  67.,  43.],
        [ 91.,  87.,  65.],
        [ 68.,  96.,  71.],
        [102.,  43.,  37.],
        [ 88., 134.,  59.]])
tensor([[ 56.,  70.],
        [ 80., 102.],
        [104., 118.],
        [ 22.,  37.],
        [118., 132.]])


In each iteration, the data loader returns one batch of data with the given batch size. If `shuffle` is set to `True`, it shuffles the training data before creating batches. Shuffling helps randomize the input to the optimization algorithm, leading to a faster reduction in the loss.

**nn.Linear**


Instead of initializing the weights & biases manually, we can define the model using the `nn.Linear` class from PyTorch, which does it automatically.

In [78]:
# Define model
model = nn.Linear(3, 2)
print(model.weight)
print(model.bias)

Parameter containing:
tensor([[ 0.4824,  0.0354,  0.5314],
        [-0.3054, -0.4184,  0.3570]], requires_grad=True)
Parameter containing:
tensor([ 0.3742, -0.2385], requires_grad=True)


PyTorch models also have a helpful `.parameters` method, which returns a list containing all the weights and bias matrices present in the model. For our linear regression model, we have one weight matrix and one bias matrix.

In [79]:
# Parameters
list(model.parameters())

[Parameter containing:
 tensor([[ 0.4824,  0.0354,  0.5314],
         [-0.3054, -0.4184,  0.3570]], requires_grad=True),
 Parameter containing:
 tensor([ 0.3742, -0.2385], requires_grad=True)]

We can use the model to generate predictions in the same way as before.



In [80]:
# Generate predictions
preds = model(inputs)
preds

tensor([[ 60.8043, -35.2123],
        [ 81.3881, -41.9987],
        [ 77.8966, -62.1657],
        [ 70.7559, -36.1677],
        [ 74.2474, -36.4863],
        [ 61.2513, -35.0993],
        [ 81.8842, -41.2233],
        [ 78.9103, -62.1141],
        [ 70.3089, -36.2808],
        [ 74.2964, -35.8240],
        [ 61.3003, -34.4370],
        [ 81.8351, -41.8856],
        [ 77.4006, -62.9411],
        [ 70.7069, -36.8300],
        [ 73.8004, -36.5993]], grad_fn=<AddmmBackward>)

**Loss Function**


Instead of defining a loss function manually, we can use the built-in loss function `mse_loss`

In [81]:
# Import nn.functional
import torch.nn.functional as F

In [82]:
# Define loss function
loss_fn = F.mse_loss

Let's compute the loss for the current predictions of our model.



In [83]:
loss = loss_fn(model(inputs), targets)
print(loss)

tensor(10396.6426, grad_fn=<MseLossBackward>)


**Optimizer**

Instead of manually manipulating the model's weights & biases using gradients, we can use the optimizer `optim.SGD`. SGD is short for "stochastic gradient descent". The term stochastic indicates that samples are selected in random batches instead of as a single group.

In [85]:
# Define optimizer
opt = torch.optim.SGD(model.parameters(), lr=1e-5)

Note that `model.parameters()` is passed as an argument to `optim.SGD` so that the optimizer knows which matrices should be modified during the update step. Also, we can specify a learning rate that controls the amount by which the parameters are modified.



**Train the model**

We are now ready to train the model. We'll follow the same process to implement gradient descent:

1. Generate predictions

2. Calculate the loss

3. Compute gradients w.r.t the weights and biases

4. Adjust the weights by subtracting a small quantity proportional to the gradient

5. Reset the gradients to zero

The only change is that we'll work batches of data instead of processing the entire training data in every iteration. Let's define a utility function fit that trains the model for a given number of epochs.

In [86]:
# Utility function to train the model
def fit(num_epochs, model, loss_fn, opt, train_dl):
    
    # Repeat for given number of epochs
    for epoch in range(num_epochs):
        
        # Train with batches of data
        for xb,yb in train_dl:
            
            # 1. Generate predictions
            pred = model(xb)
            
            # 2. Calculate loss
            loss = loss_fn(pred, yb)
            
            # 3. Compute gradients
            loss.backward()
            
            # 4. Update parameters using gradients
            opt.step()
            
            # 5. Reset the gradients to zero
            opt.zero_grad()
        
        # Print the progress
        if (epoch+1) % 10 == 0:
            print('Epoch [{}/{}], Loss: {:.4f}'.format(epoch+1, num_epochs, loss.item()))

Some things to note above:

- We use the data loader defined earlier to get batches of data for every iteration.

- Instead of updating parameters (weights and biases) manually, we use `opt.step` to perform the update and `opt.zero_grad` to reset the gradients to zero.

- We've also added a log statement that prints the loss from the last batch of data for every 10th epoch to track training progress. `loss.item` returns the actual value stored in the loss tensor.

Let's train the model for 100 epochs.

In [87]:
fit(100, model, loss_fn, opt, train_dl)

Epoch [10/100], Loss: 456.8395
Epoch [20/100], Loss: 212.9815
Epoch [30/100], Loss: 102.9979
Epoch [40/100], Loss: 60.3678
Epoch [50/100], Loss: 63.4063
Epoch [60/100], Loss: 108.2120
Epoch [70/100], Loss: 10.2574
Epoch [80/100], Loss: 33.5380
Epoch [90/100], Loss: 10.8201
Epoch [100/100], Loss: 17.4775


Let's generate predictions using our model and verify that they're close to our targets.

In [88]:
# Generate predictions
preds = model(inputs)
preds

tensor([[ 58.4371,  71.3214],
        [ 82.8471, 101.0979],
        [114.1779, 130.1587],
        [ 28.8998,  43.2596],
        [ 98.6179, 116.2060],
        [ 57.4252,  70.4051],
        [ 82.7552, 101.2251],
        [114.5592, 130.8229],
        [ 29.9118,  44.1759],
        [ 99.5380, 117.2495],
        [ 58.3452,  71.4486],
        [ 81.8352, 100.1816],
        [114.2697, 130.0315],
        [ 27.9798,  42.2162],
        [ 99.6298, 117.1223]], grad_fn=<AddmmBackward>)

In [89]:
# Compare with targets
targets

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 57.,  69.],
        [ 80., 102.],
        [118., 132.],
        [ 21.,  38.],
        [104., 118.],
        [ 57.,  69.],
        [ 82., 100.],
        [118., 134.],
        [ 20.,  38.],
        [102., 120.]])

Indeed, the predictions are quite close to our targets. We have a trained a reasonably good model to predict crop yields for apples and oranges by looking at the average temperature, rainfall, and humidity in a region. We can use it to make predictions of crop yields for new regions by passing a batch containing a single row of input.

In [90]:
model(torch.tensor([[75, 63, 44.]]))

tensor([[55.5788, 68.8894]], grad_fn=<AddmmBackward>)

The predicted yield of apples is 55.6 tons per hectare, and that of oranges is 68.9 tons per hectare.



In [91]:
model(torch.tensor([[34, 92, 67.]]))

tensor([[103.1234, 117.3753]], grad_fn=<AddmmBackward>)

**Machine Learning vs. Classical Programming**

The approach we've taken in this tutorial is very different from programming as you might know it. Usually, we write programs that take some inputs, perform some operations, and return a result.

However, in this notebook, we've defined a "model" that assumes a specific relationship between the inputs and the outputs, expressed using some unknown parameters (weights & biases). We then show the model some know inputs and outputs and train the model to come up with good values for the unknown parameters. Once trained, the model can be used to compute the outputs for new inputs.

This paradigm of programming is known as machine learning, where we use data to figure out the relationship between inputs and outputs. Deep learning is a branch of machine learning that uses matrix operations, non-linear activation functions and gradient descent to build and train models. Andrej Karpathy, the director of AI at Tesla Motors, has written a great blog post on this topics, titled Software 2.0.

This picture from book Deep Learning with Python by Francois Chollet captures the difference between classical programming and machine learning:

**Farrukh Bulbulov**